SessionSwitcher: hides all sessions on inactive display

peat-psuwit opened the pull request:
On some device, the touchscreen is still turned on even the screen is
off. To prevent the input from getting sent to the client, the session
switcher is modified to also listen to active outputs changes and hide
all sessions if all outputs are off.

Note that this PR changes the meaning of the "ActiveOutputs" property exposed on DBus.

This PR comes with a Jenkinsfile update.

Fixes ubports/ubuntu-touch#1087

peat-psuwit committed:
Screen: also calls handlers on turning on/off displays

Turning on/off display can be considered a configuration change, and
other people might want to know about that. Note that this changes the
meaning of the "ActiveOutputs" property exposed on DBus.

peat-psuwit committed:
Screen: allow multiple active outputs handlers

This allows more flexibility in notifying active outputs. The API is
updated to take a pointer as an owner key so that the it can unregister
later.

peat-psuwit committed:
SessionSwitcher: hides all sessions on inactive display

On some device, the touchscreen is still turned on even the screen is
off. To prevent the input from getting sent to the client, the session
switcher is modified to also listen to active outputs changes and hide
all sessions if all outputs are off.

Fixes https://github.com/ubports/ubuntu-touch/issues/1087

peat-psuwit committed:
Update Jenkinsfile as of ubports/build-tools@10c0987

mariogrip commented:
Does not a stop compositor event to mir turn off input events?

peat-psuwit commented:
Apparently not, otherwise we won't need this PR. See [my brain dump](https://github.com/ubports/ubuntu-touch/issues/1087#issuecomment-610830274) for more context.

UniversalSuperBox commented:
It appears that the only thing outside of USC which uses the ActiveOutputs property on D-Bus is a repowerd test. https://github.com/ubports/repowerd/blob/4d3f54cba3d5afc706ec454a224eca2d32d2158b/tests/adapter-tests/test_unity_display.cpp

peat-psuwit commented:
Turns out this PR has a side effect: it seems to hide the session(s) when lightdm is restarted. This is kind of a problem, as we should be resilient to failure.

peat-psuwit committed:
MirScreen: count active outputs in constructor

MirScreen can can be constructed after the initial_configuration()
event. Count active outputs based on current configuration so that we
have the correct info from the begining.

UniversalSuperBox commented:
Florian has tested this on the BQ E5. He experienced the original issue and did not experience it after installing this PR.

UniversalSuperBox reviewed (approved):
This additionally changes external displays. Now the external display blanks when the internal one does. Both unblank when the device gets input anyway. 

On wireless, Lomiri can restart while the internal display is off (or on, of course) and both displays will continue working after it turns back on and restarting completes. Lightdm cannot be restarted, because this ends the Aethercast session. No changes here.

On wired, either lightdm or Lomiri can be restarted at will. No changes here.

Tests fine on Qualcomm and Mediatek, merging.

mariogrip commented:
Its fine on m10 too
